internal/graph/check.go

```diff
@@ -447,56 +441,30 @@ func (cc *ConcurrentChecker) checkDirect(ctx context.Context, crc currentRequest
 	defer it.Close()
 	queryCount += 1.0
 
-	// Find the subjects over which to dispatch.
-	subjectsToDispatch := tuple.NewONRByTypeSet()
-	relationshipsBySubjectONR := mapz.NewMultiMap[string, *core.RelationTuple]()
-	hasCaveats := false
-
+	// Build the set of subjects over which to dispatch, along with metadata for
+	// mapping over caveats (if any).
+	checksToDispatch := newCheckDispatchSet()
 	for tpl := it.Next(); tpl != nil; tpl = it.Next() {
 		if it.Err() != nil {
 			return checkResultError(NewCheckFailureErr(it.Err()), emptyMetadata)
 		}
-
-		spiceerrors.DebugAssert(func() bool { return tpl.Subject.Relation != Ellipsis }, "got a terminal for a non-terminal query")
-
-		// Add the subject as an object over which to dispatch.
-		subjectsToDispatch.Add(tpl.Subject)
-		relationshipsBySubjectONR.Add(tuple.StringONR(tpl.Subject), tpl)
-		if tpl.Caveat != nil && tpl.Caveat.CaveatName != "" {
-			hasCaveats = true
-		}
+		checksToDispatch.addForRelationship(tpl)
 	}
 	it.Close()
 
-	// Convert the subjects into batched requests.
-	// To simplify the logic, +1 is added to account for the situation where
-	// the number of elements is less than the chunk size, and spare us some annoying code.
-	expectedNumberOfChunks := subjectsToDispatch.ValueLen()/int(crc.dispatchChunkSize) + 1
-	toDispatch := make([]directDispatch, 0, expectedNumberOfChunks)
-	subjectsToDispatch.ForEachType(func(rr *core.RelationReference, resourceIds []string) {
-		chunkCount := 0.0
-		slicez.ForEachChunk(resourceIds, crc.dispatchChunkSize, func(resourceIdChunk []string) {
-			chunkCount++
-			toDispatch = append(toDispatch, directDispatch{
-				resourceType: rr,
-				resourceIds:  resourceIdChunk,
-			})
-		})
-		dispatchChunkCountHistogram.Observe(chunkCount)
-	})
-
-	// If there are caveats on the incoming relationships, then we must require all results to be
-	// found, as we need to ensure that all caveats are used for building the final expression.
-	resultsSetting := crc.resultsSetting
-	if hasCaveats {
-		resultsSetting = v1.DispatchCheckRequest_REQUIRE_ALL_RESULTS
-	}
-
 	// Dispatch and map to the associated resource ID(s).
-	result := union(ctx, crc, toDispatch, func(ctx context.Context, crc currentRequestContext, dd directDispatch) CheckResult {
+	toDispatch := checksToDispatch.dispatchChunks(crc.dispatchChunkSize)
+	result := union(ctx, crc, toDispatch, func(ctx context.Context, crc currentRequestContext, dd checkDispatchChunk) CheckResult {
+		// If there are caveats on any of the incoming relationships for the subjects to dispatch, then we must require all
+		// results to be found, as we need to ensure that all caveats are used for building the final expression.
+		resultsSetting := crc.resultsSetting
+		if dd.hasIncomingCaveats {
+			resultsSetting = v1.DispatchCheckRequest_REQUIRE_ALL_RESULTS
+		}
```

Comment:
so the idea here is that because `dispatchChunks` now clusters non-ellipsis subjects relation by "caveat yes/no", we will reduce the number of dispatches that require _all_ because we've hopefully put all caveated the caveated ones in the minimum amount of chunks possible.

And this does not change with intersections and exclusions, since all dispatched elements have to be resolved.

I wonder how likely it is to find the "mixed" scenario (possible subject types with/without caveats). Because we are adding overhead by requiring sorting potentially very wide relations. Have you conducted benchmarks to understand the impact? wouldn't it perhaps be better to determine ahead of time if caveats are possible in the relation's type system? if caveats were present, the `CheckDispatchSet` will do the sorting, otherwise, it does not.

Reply:
something like `relation viewer: org#member | group#member with caveat` is likely fairly common and would be immediately improved by it. We can skip doing the sorting if no caveats are found at all (its easy to do) but then it makes the test's expected chunks inconsistent

Reply:
I guess what I'm mostly interested is that this does not regress the usual cases. If you could run a benchmark, that would help us understand the impact of the new changes.

Reply:
It can't regress the cases vs what we have now in HEAD: the previous fix did result in worse performance; this resets it as best as we can in these cases